mypyc/build.py

```diff
@@ -126,73 +101,15 @@ def get_mypy_config(paths: List[str],
     return sources, options
 
 
-shim_template_unix = """\
-#include <Python.h>
-
-PyObject *CPyInit_{full_modname}(void);
-
-PyMODINIT_FUNC
-PyInit_{modname}(void)
-{{
-    return CPyInit_{full_modname}();
-}}
-"""
-
-# As far as I could tell, Windows lacks the rpath style features we
-# would need in automatically load the shared library (located
-# relative to the module library) when a module library is loaded,
-# which means that instead we get to do it dynamically.
-#
-# We do this by, at module initialization time, finding the location
-# of the module dll and using it to compute the location of the shared
-# library. We then load the shared library with LoadLibrary, find the
-# appropriate CPyInit_ routine using GetProcAddress, and call it.
-#
-# The relative path of the shared library (from the shim library) is provided
-# as the preprocessor define MYPYC_LIBRARY.
-shim_template_windows = r"""\
+shim_template = """\
 #include <Python.h>
-#include <windows.h>
-#include <stdlib.h>
-#include <stdio.h>
-
-EXTERN_C IMAGE_DOS_HEADER __ImageBase;
-
-typedef PyObject *(__cdecl *INITPROC)();
 
 PyMODINIT_FUNC
 PyInit_{modname}(void)
 {{
-    char path[MAX_PATH];
-    char drive[MAX_PATH];
-    char directory[MAX_PATH];
-    HINSTANCE hinstLib;
-    INITPROC proc;
-
-    // get the file name of this dll
-    DWORD res = GetModuleFileName((HINSTANCE)&__ImageBase, path, sizeof(path));
-    if (res == 0 || res == sizeof(path)) {{
-        PyErr_SetString(PyExc_RuntimeError, "GetModuleFileName failed");
-        return NULL;
-    }}
-
-    // find the directory this dll is in
-    _splitpath(path, drive, directory, NULL, NULL);
-    // and use it to construct a path to the shared library
-    snprintf(path, sizeof(path), "%s%s%s", drive, directory, MYPYC_LIBRARY);
-
-    hinstLib = LoadLibrary(path);
-    if (!hinstLib) {{
-        PyErr_SetString(PyExc_RuntimeError, "LoadLibrary failed");
-        return NULL;
-    }}
-    proc = (INITPROC)GetProcAddress(hinstLib, "CPyInit_{full_modname}");
-    if (!proc) {{
-        PyErr_SetString(PyExc_RuntimeError, "GetProcAddress failed");
-        return NULL;
-    }}
-
-    return proc();
+    void *init_func = PyCapsule_Import("{libname}.{full_modname}", 0);
+    if (!init_func) return NULL;
```

Comment:
Style nit: We don't use return on the same line as 'if' in other generated code, I think. This also happens elsewhere in this PR.

Reply:
Some day I am going to pick a fight about this style guideline, because I stand by my style here pretty firmly :P, but today is not that day.

Reply:
This is probably irrelevant for _generated_ code, but there is https://www.python.org/dev/peps/pep-0007/

Reply:
My personal "house style" for C/C++ is pretty close to PEP 7 but permits the body of an if to be on the same line and unbraced if it is a "simple" special-case or error check with either a return or a goto (or maaaaybe a simple assignment) as the body. This keeps this sort of error handling code from taking up more space than it deserves, because three lines is too much.

(I guess my other gripe with PEP7 is that it doesn't cuddle the braces on both sides of an else, though we also do in mypyc.)